### PR TITLE
Update local plexapi copy, add use of pagination, add support for delta sync

### LINKED ISSUE #10: Faster (re-)synchronization
What can be done to improve (re-)synchronization after Kodi has been (re-)started? Retrieving all items and letting Kodi's media import logic handle the changeset detection and synchronization might be too slow for large libraries.

### PR DESCRIPTION
**Pull request by JeordyR**
### Main Changes
* Updated the local copy of plexapi to the latest release (3.5.0)
  * Made local patch to fix some still-existing bugs in the library
  * No need to make any import modification, so looks like we are good there
* Initially added pagination to then sync with control of the flow on our side, reverted but kept the change to the overall flow (build the ListItems in each section iteration instead of at the end), kept the container setting to make the plexapi lib paginate
  * Let me know if you want me to fully revert the flow change back to how you had it
* Updated local plexapi with ability to take a custom query string
  * Their implementation of query parameters only allows the '=' operation, this was easier than trying to rework that setup for now
* Updated import function to support fastSync
  * Currently the only check is if this is the first time syncing, then do full sync
  * Reviewed what you have going with Emby and most of it seemed emby-specific and not relevant here, but I didn't follow it all so let me know of any gaps
  * Left a comment and TODO in there to discuss what other checks we might need to do with you

### FastSync Notes

fastSync works like a charm. One of my libraries that takes about 1.5 hours to do initial import and around an hour to do synchronizations after that now finishes instantly when there are no changes, and within seconds to pick up typical changes.

The only gap I have found with the fastSync setup is that it does not detect if an item was changed from "watched" to "not watched" in plex. I couldn't find a way in the API get a delta query that would include those items. The way they store the data, the lastWatchedAt field just goes away when an item is marked unwatched and there is no other attribute that is updated to account for a change on the item.

I'm hoping we can rely on the websocket monitoring to deal with watched -> unwatched changes...?

### CODE REVIEW

**Comment by JeordyR** [comment]
Found another gap while setting it up on my main Kodi. The detection on whether or not the library has ever run is done on the import not the library section. So started out Movies with 2 libraries, they imported. Added a new one into the mix, nothing imported since it did a fastSync and there were no changes since the last one 2 minutes prior.

Will look into that tomorrow and see what I can find.

File: plexapi/compat.py
Comment (Montellese) [inline plexapi/compat.py]: Is this not needed anymore?

File: plexapi/compat.py
Comment (Montellese) [inline plexapi/compat.py:42]: Maybe keep this comment to make it obvious that this is a custom addition. But maybe replace `Montellese` with`mediaimporter.plex` ;-)

File: lib/importer.py
Comment (Montellese) [inline lib/importer.py:912]: Is this not an issue anymore?

File: lib/importer.py
Comment (Montellese) [inline lib/importer.py]: As mentioned in #10 I think we should be able to split this into multiple calls in a for loop using the proper combination of `container_size`, `maxresults` and `container_start`. Beware that using `section.totalSize` before calling `section.search()` or `section.all()` results in an additional HTTP request. So maybe it's best to access `section.totalSize` AFTER the first call to `section.search()` to determine the total number of items available.

File: lib/importer.py
Comment (Montellese) [inline lib/importer.py]: The problem is that changing certain media import specific settings should automatically lead to a full synchronization independent of whether it has already been synchronized before. For `mediaimporter.emby` I've solved this by storing a hash over the settings affecting synchronization in the settings and before every sync I compare the stored hash with a newly calculated hash to see if anything has changed (see https://github.com/Montellese/mediaimporter.emby/commit/661251503cd660e7f64366ac09322f005132e055).

Furthermore I've added a "force full synchronization" option to the settings to let the user manually force a full synchronization (see https://github.com/Montellese/mediaimporter.emby/commit/13cf89223ff53968068479560eba49294adf9aaf).

File: lib/importer.py
Comment (Montellese) [inline lib/importer.py]: Is it necessary to do this check in both directions? Wouldn't it be enough to just add all `updatedPlexItems` to `plexItems` and then only extend `plexItems` with the `watchedPlexItems` which aren't already part of `plexItems`?

File: lib/importer.py
Comment (Montellese) [inline lib/importer.py]: `lastViewedAt` can be passed into `Section.search()` as a kwargs, see e.g. https://github.com/Montellese/mediaimporter.plex/blob/9bd0b05422d1dd49cc78ba7f2a492dd73a74cc5e/plexapi/library.py#L765
Unfortunately for tvshows it seems to be `show.lastViewedAt`, see https://github.com/Montellese/mediaimporter.plex/blob/9bd0b05422d1dd49cc78ba7f2a492dd73a74cc5e/plexapi/library.py#L826 and for episodes `episode.lastViewedAt`, see https://github.com/Montellese/mediaimporter.plex/blob/9bd0b05422d1dd49cc78ba7f2a492dd73a74cc5e/plexapi/library.py#L829

Concerning `updatedAt` I didn't find any filters but maybe we can ask `python-plexapi` to add them as well then we don't need any extra patches on the library.

**Review by Montellese — CHANGES_REQUESTED** [review]
Thanks for working on this. I've added some comments and maybe we have to contact the maintainers of `python-plexapi` (again).

**Comment by Montellese** [comment]
> Found another gap while setting it up on my main Kodi. The detection on whether or not the library has ever run is done on the import not the library section. So started out Movies with 2 libraries, they imported. Added a new one into the mix, nothing imported since it did a fastSync and there were no changes since the last one 2 minutes prior.
> 
> Will look into that tomorrow and see what I can find.

See my comment in the review concerning detecting changes to the media import's settings. That would solve this issue.

File: lib/importer.py
Comment (JeordyR) [reply lib/importer.py]: Looking through the implementation the arguments always get joined with an '=' operation:

https://github.com/Montellese/mediaimporter.plex/blob/35c0575e5a38dfc99d78a91aeeff429a262c52d9/plexapi/utils.py#L93

So we would end up with lastViewedAt=<time> instead of lastViewedAt>=<time>

I didn't see anywhere that their current implementation allows specifying the operator.

I threw the custom_filter_string option in real quick just to get going. You think it would be better to rework their filter collection into allowing operator specification (and standardization between the types maybe)?

File: lib/importer.py
Comment (JeordyR) [reply lib/importer.py]: That makes more sense, will update.

File: lib/importer.py
Comment (JeordyR) [reply lib/importer.py]: Ahh I see, that's the part I was missing. Will get that updated.

File: lib/importer.py
Comment (JeordyR) [reply lib/importer.py:912]: I didnt' see any code in or used by Api.toFileItem that makes calls back out to Plex that could raise BadRequest.

It is still up around the .search calls

File: lib/importer.py
Comment (JeordyR) [reply lib/importer.py]: This did end up working perfectly, thanks!

**Comment by JeordyR** [comment]
> Thanks for working on this. I've added some comments and maybe we have to contact the maintainers of `python-plexapi` (again).

No problem, has been a fun project. Once we get everything fully sorted with the local patches I'll do a PR to the plexapi folks. I'm hoping we get get to the point of directly using the library and making a script addon out of it to import instead of keeping a local copy, but we'll see.

For those compatibility things I want to see if I can find a way to detect when running on Kodi so we can commit those up wrapped in a check as well.

File: lib/importer.py
Comment (JeordyR) [reply lib/importer.py]: Got both of these added and tested

**Comment by JeordyR** [comment]
@Montellese Finished updating based on comments with the settings change detection, force sync option in settings, plexapi compat items, and pagination fix. Left 2 of the conversations open for feedback. Let me know if you have any other notes or want the commits squashed any more.

File: lib/importer.py
Comment (Montellese) [reply lib/importer.py]: They already seem to support `!=` as an operator. The approach seems to be that you e.g. specify `{ "country!": "Italy" }` i.e. you add the first part of the operator to the end of the filter name / category. At least that what it looks like when I look at https://github.com/Montellese/mediaimporter.plex/blob/9bd0b05422d1dd49cc78ba7f2a492dd73a74cc5e/plexapi/library.py#L628-L630
Maybe that approach can be extended with `<` and `>` as well. Then we only need to have `updatedAt` as a valid filter and we should be good to go.

BTW while looking through the filters I also saw `addedAt`. What is the value of `updatedAt` for newly added items? Is it the same value `addedAt` or is it empty? If the latter applies we would also have to retrieve based on `addedAt`.

File: lib/importer.py
Comment (Montellese) [reply lib/importer.py:912]: TBH I'm not sure. The comment below seems to indicate that the issue is with `Api.convertDateTimeToDbDateTime` but that sounds odd.

File: lib/importer.py
Comment (Montellese) [inline lib/importer.py:913]: Shouldn't this be an `elif` or at least check if `fastSync` is already `False`? There's no reason to check this if `lastSync` is not set.

Comment (JeordyR) [reply lib/importer.py:913]: I had made it a separate if intentionally so that on first import of a library the hash still gets generated. I'll put it back to an elif and throw a call to calculate hash under the first run check.

File: lib/importer.py
Comment (JeordyR) [reply lib/importer.py]: Ah I see, I didn't process that at all since variable names can't have operators. I'm guessing it ends up working since it stays in kwargs. Will probly have either pass an unpacked dict or a kwargs dict into it, not standard but should probly work to still fill **kwargs with the contents of kwargs instead of kwargs['kwargs'] = {}, I'll play around with it.

updatedAt ends up populated with almost the same value as addedAt. In Plex when you add a new item, addedAt is populated without updatedAt but it immediately performs an update on the item and updatedAt get populated. It is possible for us to be running an import within that few-second window between add and update and miss an item but I'm not sure the chance of it happening is high enough to warrant an additional API call, what do you think?

File: lib/importer.py
Comment (JeordyR) [reply lib/importer.py:912]: That comment doesn't seem to describe a possible situation, but I'll just throw the try/catch back around it to be safe.

File: lib/importer.py
Comment (JeordyR) [reply lib/importer.py]: Got it working with some patches to plexapi. Passing an unpacked dict works, kwargs= does not.

Lmk what you think on above with addedAt and if you're good I'll get the commits squashed down.

Comment (Montellese) [reply lib/importer.py]: Let's go with what you got and if we notice any issues we can still add the call for `addedAt`.

File: lib/importer.py
Comment (Montellese) [inline lib/importer.py:921]:
```suggestion
        updatedFilter = {"updatedAt>": lastSyncEpoch}
        watchedFilter = {"lastViewedAt>": lastSyncEpoch}
```

File: lib/importer.py
Comment (Montellese) [inline lib/importer.py]: I would minimize the code envelopped by this `try-except` to the following part:
```python
item = Api.toFileItem(plexServer, plexItem, mediaType, plexLibType)
if not item:
    continue

itemsToImport.append(item)
```

File: plexapi/library.py
Comment (Montellese) [inline plexapi/library.py:492]: Reading this documentation it basically states that `updatedAt>=XYZ&lastViewedAt>=XYZ` should result in all items being returned which where either watched or updated since `XYZ`. But IIRC you stated that it doesn't work. Did you try it?

File: lib/importer.py
Comment (Montellese) [reply lib/importer.py:912]: This was added by @enen92 in 6337f684b0301c5ee5a6a66e88a5bac643797922. Maybe he can provide some details on this question.

**Comment by Montellese** [comment]
Looks much better. It would be great if you could propose the `updatedAt` change for `ALLOWED_FILTERS` and the `>` and `<` change upstream and see what they think about it.

File: lib/importer.py
Comment (enen92) [reply lib/importer.py:912]: Yeah this was added because a single item (broken cache in your Plex server) could break the whole import process when you request details for the item. When this happened (it was Ralph breaks the internet in one of the test servers) this was the exception that was raised (result of 404) when querying the movie id.

Comment (enen92) [reply lib/importer.py:912]: So if Api.tofileitem() still results in a call to the server, it's better to leave the exception there. The failure is rare but it can happen when you have a huge library; having it crash in those cases is not good as the import process takes a long time (at least on the first import)

File: lib/importer.py
Comment (JeordyR) [reply lib/importer.py]: Gotchya

File: plexapi/library.py
Comment (JeordyR) [reply plexapi/library.py:492]: I didn't try it using their library, but tried it with direct API calls and it AND'd them together not OR'd them.

Also in [these docs](https://github.com/Arcanemagus/plex-api/wiki/Plex-Web-API-Overview) of the api it calls out `When multiple parameters are passed, they are logically ANDed together`

File: lib/importer.py
Comment (JeordyR) [reply lib/importer.py:921]: What is your thought for this? Unofficial convention is single quotes for dictionary keys and all other dicts I found in the code (self._state in player.py for example) use single quotes for their keys.

**Comment by JeordyR** [comment]
> Looks much better. It would be great if you could propose the `updatedAt` change for `ALLOWED_FILTERS` and the `>` and `<` change upstream and see what they think about it.

Definitely will. Can you think of any way off the top of your head to detect that code is being executed in Kodi? I'd like to also get the compatibility changes committed upstream with specific check for Kodi so we can just use upstream directly. Plan to dig into it and see if I can find a check to use, but lmk if you can think of anything off-hand.

File: lib/importer.py
Comment (Montellese) [reply lib/importer.py:921]: I'm not a proficient python developer to have an actual opinion on this ;-) I thought I used double quotes everywhere but for no other reason than consistency. If there is a convention I'm all for following it.

**Comment by Montellese** [comment]
> > Looks much better. It would be great if you could propose the `updatedAt` change for `ALLOWED_FILTERS` and the `>` and `<` change upstream and see what they think about it.
> 
> Definitely will. Can you think of any way off the top of your head to detect that code is being executed in Kodi? I'd like to also get the compatibility changes committed upstream with specific check for Kodi so we can just use upstream directly. Plan to dig into it and see if I can find a check to use, but lmk if you can think of anything off-hand.

I'm pretty sure upstream doesn't want to have any Kodi-specific code in their library. Maybe there is a Pythonic way to do this in our code with some kind of wrapper?

File: lib/importer.py
Comment (JeordyR) [reply lib/importer.py:921]: It's pretty all over the place with python and varies by developer and use of the two isn't defined in PEP8 to clear it up.

The main convention I have seen is single-quote's for object-y references (dict keys mostly), singles when passing a short string argument in, and double for pretty much everything else so that single's can be used inside them. With triple-doubles for block comments and docstrings, never triple-singles.

I think all the dicts that I found were commits by enen. Throughout the file there is a mix of single and double for the long strings (probly a mix of you and enen :P)

**Comment by JeordyR** [comment]
> > > Looks much better. It would be great if you could propose the `updatedAt` change for `ALLOWED_FILTERS` and the `>` and `<` change upstream and see what they think about it.
> > 
> > 
> > Definitely will. Can you think of any way off the top of your head to detect that code is being executed in Kodi? I'd like to also get the compatibility changes committed upstream with specific check for Kodi so we can just use upstream directly. Plan to dig into it and see if I can find a check to use, but lmk if you can think of anything off-hand.
> 
> I'm pretty sure upstream doesn't want to have any Kodi-specific code in their library. Maybe there is a Pythonic way to do this in our code with some kind of wrapper?

We might be able to override their compat module with something like this:

```
import plexapi
plexapi.compat = lib.compat
```

I'm not sure if things that were initialized on the initial `import plexapi` that may have already imported something from compat would actually end up pointing at our version or not though.

I'll do some testing and do a PR of that if I can come up with something that works after getting the other changes merged upstream.

**Review by Montellese — APPROVED** [review]
Looks good, thanks a lot. Can you cleanup the commits a bit?

**Comment by JeordyR** [comment]
> Looks good, thanks a lot. Can you cleanup the commits a bit?

@Montellese 

Got all of the delta sync and pagination changes merged down into 1 commit and left the plexapi update to 3.5.0 separate for the update call-out. Let me know if you want them renamed or anything.

File: lib/importer.py
Comment (Montellese) [inline lib/importer.py:942]: Shouldn't all of these changes to `lib/importer.py` be part of the second commit? They don't have anything to do with the update of `plexapi` right?

**Review by Montellese — CHANGES_REQUESTED** [review]
For some reason I can't find the addition of `updatedAt` in `ALLOWED_FILTERS` anymore. Has it gone missing or am I blind?

**Comment by JeordyR** [comment]
> For some reason I can't find the addition of `updatedAt` in `ALLOWED_FILTERS` anymore. Has it gone missing or am I blind?

Yeah, you're right. I'm not sure what happened, seemed to have lost some commits in the squashing and merging. I'll get the branch reset back to master and apply the changes back in to clean commits this afternoon.

**Comment by JeordyR** [comment]
@Montellese Should be all situated. All of the plexapi changes are together in the "update + local patches" and all of the others are in a single commit.

Doing some more testing now to make sure everything is still working as expected and nothing else was lost, will post another update on that in a bit.

File: plex/provider_observer.py
Comment (Montellese) [inline plex/provider_observer.py:383]: Why did you change this? `pylint` and `flake8` both usually propose to use `x in (foo, bar)` instead of `x == foo or x == bar`.

File: plex/server.py
Comment (Montellese) [inline plex/server.py:38]: Is this really necessary?

Comment (JeordyR) [reply plex/server.py:38]: I can put it back, just used to `\` being considered bad form and preferring to do pretty much any other multi-line solution other than `\` and changed it out of habit. Lmk if you want that changed back.

Comment (Montellese) [reply plex/server.py:38]: Ah ok. It just looks redundant to me but if it's better code style I can live with it.

File: plex/provider_observer.py
Comment (JeordyR) [reply plex/provider_observer.py:383]: Oh, I just saw the `\` and merged over-top. Pushing bringing in another commit now to set it to

```
            if changesetType in (
                    xbmcmediaimport.MediaImportChangesetTypeAdded,
                    xbmcmediaimport.MediaImportChangesetTypeChanged
            ):
```

Comment (JeordyR) [reply plex/provider_observer.py:383]: Updated

**Comment by JeordyR** [comment]
> @Montellese Should be all situated. All of the plexapi changes are together in the "update + local patches" and all of the others are in a single commit.
> 
> Doing some more testing now to make sure everything is still working as expected and nothing else was lost, will post another update on that in a bit.

Tests done, looks good on both Movie and TV libraries.

**Review by Montellese — APPROVED** [review]
Thanks a lot for working on this and working through all my comments!

**Comment by Montellese** [comment]
I gave this a try and there seems to be something off with the progress indication when using pagination. When retrieving a list of hundreds of movies or episodes the progress indication starts at 0% goes up to 100%, then back to 0% from where it goes up to 100% and so on until all movies / episodes have been retrieved.

EDIT: The problem only occurs during retrieval phase. Processing / Synchronization phase (within Kodi) is fine.

**Comment by JeordyR** [comment]
@Montellese 

I'm guessing the iteration on the plexItems:
https://github.com/Montellese/mediaimporter.plex/blob/48102edd6e04eef2dc312a6f3acf56ba15aa0998/lib/importer.py#L1006-L1009

 is overriding the progress indication for the outer section iteration:
https://github.com/Montellese/mediaimporter.plex/blob/48102edd6e04eef2dc312a6f3acf56ba15aa0998/lib/importer.py#L958-L964

With, say 1200 movies, it would report like this:
* 0/100 (initial section iteration before search)
* 1-100/100 (plexItems iteration)
* 100/1200 (back up to section iteration, sectionProgressTotal now populated correctly after search)
* 1-100/100 (plexItems iteration again)
* 200/1200 (back up to section iteration)
...
 
and the plexItems iteration takes much longer than the section data pull, so it probly flashes past the 100/1200 etc. too fast to really notice.

Maybe reworking the plexItem iteration to just also reference sectionProgress and sectionProgressTotal instead of its own values would work? sectionProgress is iterated within that loop already anyways.

**Comment by Montellese** [comment]
Yeah https://github.com/Montellese/mediaimporter.plex/blob/48102edd6e04eef2dc312a6f3acf56ba15aa0998/lib/importer.py#L1008 should probably use `sectionProgress` and `sectionProgressTotal` as well.

**Comment by Montellese** [comment]
I just ran into the following python exception when `mediaimporter.plex` was automatically running after I started Kodi:
```
EXCEPTION Thrown (PythonToCppException) : -->Python callback/script returned the following error<--
 - NOTE: IGNORING THIS CAN LEAD TO MEMORY LEAKS!
Error Type: <class 'ValueError'>
Error Contents: Invalid format string
Traceback (most recent call last):
  File "D:\Coding\Cpp\xbmc.montellese\kodi-build.win10\addons\mediaimporter.plex\importer.py", line 20, in <module>
    importer.run(sys.argv)
  File "D:\Coding\Cpp\xbmc.montellese\kodi-build.win10\addons\mediaimporter.plex\lib\importer.py", line 1181, in run
    actionMethod(handle, options)
  File "D:\Coding\Cpp\xbmc.montellese\kodi-build.win10\addons\mediaimporter.plex\lib\importer.py", line 923, in execImport
    lastSyncEpoch = parser.parse(lastSync).strftime('%s')
ValueError: Invalid format string
-->End of Python script error report<--
```
All my Plex related media imports have already been imported into the Kodi library and therefore should have a valid `lastSync` value.

Manually starting the import process on both my Plex related media providers causes the same problem to appear.

**Comment by Montellese** [comment]
I just checked the documentation of `datetime.strftime()`, see https://docs.python.org/3/library/datetime.html#strftime-strptime-behavior, and there's no `%s` format string.

**Comment by Montellese** [comment]
I have fixed both issues.

**Comment by Montellese** [comment]
@JeordyR for some reason fast / delta sync doesn't seem to be working anymore. I'm not sure if anything changed in the PMS backend but the `updatedAt` property of a movie doesn't change when I e.g. change the title of the movie. It only seems to update when all the metadata of the movie is manually refreshed. My current PMS version is `1.23.1.4571`. Can you check at your end if it is still working for you? I also tried manual requests with the same results.

**Pull request merged**
